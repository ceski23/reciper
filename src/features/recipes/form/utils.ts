import { group } from 'radash'
import { type RecipeFormValues } from 'features/recipes/form/scheme'
import { type Recipe } from 'features/recipes/types'

export const mapFormValuesToRecipe = (values: RecipeFormValues) => ({
	...values,
	ingredients: values.ingredients.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	instructions: values.instructions.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	tags: values.tags.map(tag => tag.text),
} satisfies Omit<Recipe, 'id' | 'addedDate'>)

export const mapRecipeToFormValues = (recipe: Recipe) => ({
	...recipe,
	ingredients: Object.entries(group(recipe.ingredients, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	instructions: Object.entries(group(recipe.instructions, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	tags: recipe.tags.map(tag => ({ text: tag })),
} satisfies RecipeFormValues)
