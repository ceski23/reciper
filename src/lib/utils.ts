export const isDefined = <T>(item: T): item is NonNullable<T> => item !== undefined && item !== null
