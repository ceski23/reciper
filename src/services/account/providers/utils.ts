export const randomBytes = (size: number) => {
  return crypto.getRandomValues(new Uint8Array(size))
}

export const base64url = (bytes: any) => {
  return btoa(String.fromCharCode(...bytes))
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
}

export const generateCodeChallenge = async (code_verifier: string) => {
  const codeVerifierBytes = new TextEncoder().encode(code_verifier)
  const hashBuffer = await crypto.subtle.digest('SHA-256', codeVerifierBytes)
  return base64url(new Uint8Array(hashBuffer))
}