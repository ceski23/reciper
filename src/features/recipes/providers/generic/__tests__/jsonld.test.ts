import { describe, expect, it } from 'vitest'
import { createRecipe } from 'features/recipes/providers/websites/__tests__/utils'
import { extractJsonLD } from '../jsonld'
import { recipe1, recipe2, recipe3, recipe4, recipe5 } from './jsonld.fixtures'

describe('should scrape recipes using LD+JSON format', () => {
	const scrapeRecipe = async (data: string) => extractJsonLD(new DOMParser().parseFromString(data, 'text/html'))

	it('should scrape valid recipe for Shepherdâ€™s Pie', async () => {
		const partialRecipe = await scrapeRecipe(recipe1)

		expect(partialRecipe).toMatchSnapshot()
		expect(createRecipe(partialRecipe)).toBeValidRecipe()
	})

	it('should scrape valid recipe for Homemade Buttery Flaky Pie Crust', async () => {
		const partialRecipe = await scrapeRecipe(recipe2)

		expect(partialRecipe).toMatchSnapshot()
		expect(createRecipe(partialRecipe)).toBeValidRecipe()
	})

	it('should scrape valid recipe for Pecan Pie', async () => {
		const partialRecipe = await scrapeRecipe(recipe3)

		expect(partialRecipe).toMatchSnapshot()
		expect(createRecipe(partialRecipe)).toBeValidRecipe()
	})

	it('should scrape valid recipe for Apple Cranberry Pie', async () => {
		const partialRecipe = await scrapeRecipe(recipe4)

		expect(partialRecipe).toMatchSnapshot()
		expect(createRecipe(partialRecipe)).toBeValidRecipe()
	})

	it('should scrape valid recipe for Creamy Chicken Casserole', async () => {
		const partialRecipe = await scrapeRecipe(recipe5)

		expect(partialRecipe).toMatchSnapshot()
		expect(createRecipe(partialRecipe)).toBeValidRecipe()
	})

	it('should not found recipe', async () => {
		await expect(scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Nothing"
            }
          </script>
        `)).rejects.toThrow('No JSON-LD data available')
	})

	it('should detect recipe\'s title', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "name": "Test recipe"
            }
          </script>
        `)

		expect(recipe.name).toBe('Test recipe')
	})

	it('should detect recipe\'s image', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": "test.jpg"
                  }
                </script>
              `)

			expect(recipe.image).toBe('test.jpg')
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": ["test.jpg", "test2.png"]
                  }
                </script>
              `)

			expect(recipe.image).toBe('test.jpg')
		}
	})

	it('should detect recipe\'s description', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "description": "Testing test"
            }
          </script>
        `)

		expect(recipe.description).toBe('Testing test')
	})

	it('should detect recipe\'s instructions', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "recipeInstructions": [
                      { "@type": "HowToStep", "text": "Test1" },
                      { "@type": "HowToStep", "text": "Test2" },
                      { "@type": "HowToStep", "text": "Test3" }
                    ]
                  }
                </script>
              `)

			expect(recipe.instructions).toStrictEqual([
				{ text: 'Test1' },
				{ text: 'Test2' },
				{ text: 'Test3' },
			])
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "recipeInstructions": {
                      "@type": "HowToSection",
                      "steps": "Test1"
                    }
                  }
                </script>
              `)

			expect(recipe.instructions).toStrictEqual([
				{ text: 'Test1' },
			])
		}
		// TODO: test for another forms of instructions (HowToStep etc.)
	})

	it('should detect recipe\'s preparation time', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "prepTime": "PT10M"
            }
          </script>
        `)

		expect(recipe.prepTime).toBe(10)
	})

	it('should detect recipe\'s servings', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "recipeYield": "1 loaf"
            }
          </script>
        `)

		expect(recipe.servings).toBe(1)
	})

	it('should detect recipe\'s rating', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "aggregateRating": {
                "@type": "AggregateRating",
                "ratingValue": "4"
              }
            }
          </script>
        `)

		expect(recipe.rating).toBe(4)
	})

	it('should detect recipe\'s calories', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "nutrition": {
                "@type": "NutritionInformation",
                "calories": "240 calories"
              }
            }
          </script>
        `)

		expect(recipe.calories).toBe(240)
	})

	it('should detect recipe\'s ingredients', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "recipeIngredient": [
                "Test1",
                "Test2"
              ]
            }
          </script>
        `)

		expect(recipe.ingredients).toStrictEqual([
			{ text: 'Test1' },
			{ text: 'Test2' },
		])
	})

	it('should detect recipe\'s tags', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <script type="application/ld+json">
            {
              "@context": "https://schema.org/",
              "@type": "Recipe",
              "recipeCategory": "Test1",
              "recipeCuisine": "Test2"
            }
          </script>
        `)

		expect(recipe.tags).toStrictEqual(['test1', 'test2'])
	})

	it('should detect recipe\'s images', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": "test.jpg"
                  }
                </script>
              `)

			expect(recipe.gallery).toStrictEqual(['test.jpg'])
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": {
                      "@type": "ImageObject",
                      "url": "test.jpg",
                      "width": 1,
                      "height": 1
                    }
                  }
                </script>
              `)

			expect(recipe.gallery).toStrictEqual(['test.jpg'])
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": [
                      {
                        "@type": "ImageObject",
                        "url": "test1.jpg",
                        "width": 1,
                        "height": 1
                      },
                      {
                        "@type": "ImageObject",
                        "url": "test2.jpg",
                        "width": 1,
                        "height": 1
                      }
                    ]
                  }
                </script>
              `)

			expect(recipe.gallery).toStrictEqual(['test1.jpg', 'test2.jpg'])
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <script type="application/ld+json">
                  {
                    "@context": "https://schema.org/",
                    "@type": "Recipe",
                    "image": ["test1.jpg", "test2.jpg"]
                  }
                </script>
              `)

			expect(recipe.gallery).toStrictEqual(['test1.jpg', 'test2.jpg'])
		}
	})
})
