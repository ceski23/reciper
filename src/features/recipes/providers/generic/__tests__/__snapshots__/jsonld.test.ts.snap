// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Apple Cranberry Pie 1`] = `
{
  "calories": 357,
  "color": "#ffffff",
  "description": "Buttery, flaky pie crust with spiced apples, tart cranberries, and creamy vanilla ice cream. This is the perfect holiday dessert!",
  "gallery": [
    "https://www.twopeasandtheirpod.com/wp-content/uploads/2022/11/Apple-Cranberry-Pie-43.jpg",
    "https://www.twopeasandtheirpod.com/wp-content/uploads/2022/11/Apple-Cranberry-Pie-43-500x500.jpg",
    "https://www.twopeasandtheirpod.com/wp-content/uploads/2022/11/Apple-Cranberry-Pie-43-500x375.jpg",
    "https://www.twopeasandtheirpod.com/wp-content/uploads/2022/11/Apple-Cranberry-Pie-43-480x270.jpg",
  ],
  "image": "https://www.twopeasandtheirpod.com/wp-content/uploads/2022/11/Apple-Cranberry-Pie-43.jpg",
  "ingredients": [
    {
      "text": "2 1/2 cups all-purpose flour",
    },
    {
      "text": "1 tablespoon sugar",
    },
    {
      "text": "1 teaspoon salt",
    },
    {
      "text": "1 cup cold unsalted butter, (cut into cubes)",
    },
    {
      "text": "1/2 cup cold buttermilk",
    },
    {
      "text": "1-2 tablespoons cold water",
    },
    {
      "text": "1  large egg beaten with 1 tablespoon water, (for egg wash)",
    },
    {
      "text": "Turbinado sugar, (optional)",
    },
    {
      "text": "4  large Honeycrisp apples, ( cored, peeled, and sliced (about 6-7 cups))",
    },
    {
      "text": "Juice of 1/2 lemon",
    },
    {
      "text": "1 1/2 cups cranberries, (fresh or frozen)",
    },
    {
      "text": "1 cup granulated sugar",
    },
    {
      "text": "1/3 cup all-purpose flour",
    },
    {
      "text": "1 teaspoon ground cinnamon",
    },
    {
      "text": "2 teaspoons  orange zest",
    },
    {
      "text": "1 teaspoon vanilla extract",
    },
  ],
  "instructions": [
    {
      "text": "First, make the pie crust. Combine the flour, sugar, and salt in a large bowl. Add the cubed butter and toss to coat. Dump the mixture out onto a clean surface and use a rolling pin to roll the butter into thin sheets, combining it with the flour. Use a bench scraper to scrape the rolling pin and to bring the mixture back into a pile as necessary. Continue until all of the butter is incorporated into the flour. Mixture will be very flaky. Return mixture to the bowl and place in the freezer for 15 minutes to chill the butter.",
    },
    {
      "text": "Remove from freezer and add the buttermilk. Use a spoon and then your hands to stir the mixture until it comes together into a ball. If mixture is too dry, add the water a tablespoon at a time. Divide the dough in two and flatten into disks. Wrap each disk in plastic wrap and chill in the fridge while you make the filling.",
    },
    {
      "text": "Put the apple slices in a large bowl and drizzle with fresh lemon juice. Toss to coat the apples. This will keep the apples from browning. Add the cranberries, sugar, flour, cinnamon, orange zest, and vanilla extract. Stir until combined.",
    },
    {
      "text": "Preheat oven to 375°F.",
    },
    {
      "text": "Remove one disk of pie dough from the refrigerator. On a lightly floured surface, use a rolling pin to roll out the dough into a 12-inch circle. Dough should be about 1/4 inch thick. Transfer dough to a 9-inch pie pan. Pat with your fingers, making sure it is smooth. Trim the extra overhang of dough with a knife and set aside. Fill pie with apple cranberry filling. Place the pie in the fridge while you roll out the second disk of dough.",
    },
    {
      "text": "Roll our the dough and use a pastry wheel, pizza cutter, or sharp knife to cut the pie dough into strips, about 1-2 inches wide. Carefully weave the pie dough strips over and under one another. Press the edges of the dough strips into the bottom pie crust edges to seal. Trim off the excess dough and pinch the edges using your fingers to create a pretty edge. You can also use a fork to seal the edges. Using a pastry brush, lightly brush crust and edges with egg wash. Sprinkle pie with turbinado sugar, if using.",
    },
    {
      "text": "Place the pie on a foil-lined baking sheet, just in case the filling bubbles out.",
    },
    {
      "text": "Bake for 60-80 minutes until the crust is golden brown and filling is bubbling. If the top of the pie is browning too quickly, you can tent the top with aluminum foil or use a pie crust shield. Make sure you check it halfway through to be safe.",
    },
    {
      "text": "When the pie is done baking, remove from oven and cool completely before serving so it can set up, about 3 hours.",
    },
  ],
  "name": "Apple Cranberry Pie",
  "prepTime": 105,
  "rating": 4.71,
  "servings": 12,
  "tags": [
    "dessert",
    "american",
    "apple",
    "cranberry",
    "pie",
  ],
}
`;

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Creamy Chicken Casserole 1`] = `
{
  "calories": 340,
  "color": "#ffffff",
  "description": "The chicken in this chicken casserole is incredibly tender - you will savor every morsel! This chicken bake re-heats really well (if you are lucky enough to have leftovers).",
  "gallery": [
    "https://natashaskitchen.com/wp-content/uploads/2022/02/Chicken-Caserole-SQ.jpg",
    "https://natashaskitchen.com/wp-content/uploads/2022/02/Chicken-Caserole-SQ-500x500.jpg",
    "https://natashaskitchen.com/wp-content/uploads/2022/02/Chicken-Caserole-SQ-500x375.jpg",
    "https://natashaskitchen.com/wp-content/uploads/2022/02/Chicken-Caserole-SQ-480x270.jpg",
  ],
  "image": "https://natashaskitchen.com/wp-content/uploads/2022/02/Chicken-Caserole-SQ.jpg",
  "ingredients": [
    {
      "text": "2 1/2 lbs chicken breasts  ((4 to 5 large), cut into 1 -inch thick strips)",
    },
    {
      "text": "1 1/2 tsp sea salt  (plus more to taste)",
    },
    {
      "text": "1/4 tsp  black pepper (plus more to taste)",
    },
    {
      "text": "1 cup all-purpose flour  (to coat the chicken)",
    },
    {
      "text": "6 Tbsp olive oil (divided)",
    },
    {
      "text": "1 pound white mushrooms (thickly sliced)",
    },
    {
      "text": "1  medium onion (finely diced)",
    },
    {
      "text": "3  garlic cloves (minced)",
    },
    {
      "text": "3 Tbsp unsalted butter",
    },
    {
      "text": "3 Tbsp all-purpose flour (for the sauce)",
    },
    {
      "text": "1 1/2 cups chicken broth",
    },
    {
      "text": "1 Tbsp lemon juice",
    },
    {
      "text": "1 cup half and half (or 1/2 cup milk + 1/2 cup heavy cream)",
    },
  ],
  "instructions": [
    {
      "text": "Season chicken with salt and pepper and dredge both sides in flour. Heat 3 Tbsp oil in a large skillet over medium/high heat. Once oil is hot, add chicken and saute until golden. Work in batches if needed so you don't crowd your pan. Don't worry about cooking the chicken through; it still needs to go in the oven. Transfer chicken to a 13x9 casserole baking dish.",
    },
    {
      "text": "Scrape of any debris from the skillet. Add 3 Tbsp oil along with sliced mushrooms and diced onions and saute until soft and golden. Add garlic and saute another 1-2 min. Spread mixture over the chicken.",
    },
    {
      "text": "In a medium saucepan (or in the same pan you've been using), melt 3 Tbsp butter. Whisk in 3 Tbsp flour until lightly golden (1-2 min).",
    },
    {
      "text": "Add 1 1/2 cups chicken broth, 1 Tbsp lemon juice, 1/4 tsp pepper, and whisk until smooth. Add 1 cup half & half and bring to a simmer. Season generously with salt and pepper to taste.",
    },
    {
      "text": "Pour sauce over the chicken and mushrooms. Cover with foil and bake at 350˚F for 45 minutes. Serve warm or hot over mashed potatoes, pasta or rice.",
    },
  ],
  "name": "Chicken Casserole Recipe (Chicken Gloria)",
  "prepTime": 75,
  "rating": 4.95,
  "servings": 10,
  "tags": [
    "main course",
    "american",
    "chicken casserole",
  ],
}
`;

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Homemade Buttery Flaky Pie Crust 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "This recipe is enough for a double crust pie. If you only need 1 crust for your pie, freeze the other half per the Freezing Instructions below. Is your pie dough tearing, cracking, or crumbling as you try to roll it out? See recipe Notes.",
  "gallery": [
    "https://sallysbakingaddiction.com/wp-content/uploads/2015/07/perfect-pie-crust-3-225x225.jpg",
    "https://sallysbakingaddiction.com/wp-content/uploads/2015/07/perfect-pie-crust-3-260x195.jpg",
    "https://sallysbakingaddiction.com/wp-content/uploads/2015/07/perfect-pie-crust-3-320x180.jpg",
    "https://sallysbakingaddiction.com/wp-content/uploads/2015/07/perfect-pie-crust-3.jpg",
  ],
  "image": "https://sallysbakingaddiction.com/wp-content/uploads/2015/07/perfect-pie-crust-3-225x225.jpg",
  "ingredients": [
    {
      "text": "2 and 1/2 cups (315g) all-purpose flour (spooned & leveled), plus more for shaping and rolling",
    },
    {
      "text": "1 teaspoon salt",
    },
    {
      "text": "6 Tablespoons (85g) unsalted butter, chilled and cubed",
    },
    {
      "text": "2/3 cup (130g) vegetable shortening, chilled",
    },
    {
      "text": "1/2 cup (120ml) ice cold water",
    },
  ],
  "instructions": [
    {
      "text": "Whisk the flour and salt together in a large bowl.",
    },
    {
      "text": "Add the butter and shortening. Using a pastry cutter or two forks, cut the butter and shortening into the mixture until it resembles coarse meal (pea-sized bits with a few larger bits of fat is OK). In this step, you’re only breaking up the cold fat into tiny little flour-coated pieces; you’re not completely incorporating it. Do not overwork the ingredients.",
    },
    {
      "text": "Measure 1/2 cup (120ml) of water in a cup. Add ice. Stir it around. From that, measure 1/2 cup (120ml) of water, since the ice has melted a bit. Drizzle the cold water in, 1 Tablespoon (15ml) at a time, and stir with a rubber spatula or wooden spoon after every Tablespoon has been added. Stop adding water when the dough begins to form large clumps. I always use about 1/2 cup of water, and need a little more in dry winter months. Do not add any more water than you need.",
    },
    {
      "text": "Transfer the pie dough to a floured work surface. Using floured hands, fold the dough into itself until the flour is fully incorporated into the fats. The dough should come together easily and should not feel overly sticky. Avoid overworking the dough. If it feels a bit too dry or crumbly, dip your fingers in the ice water and then continue bringing dough together with your hands. If it feels too sticky, sprinkle on more flour and then continue bringing dough together with your hands. Form it into a ball. Use a sharp knife to cut it in half. If it’s helpful, you should have about 1 lb, 8 ounces dough total (about 680g). Gently flatten each half into 1-inch-thick discs using your hands.",
    },
    {
      "text": "Wrap each tightly in plastic wrap. Refrigerate for at least 2 hours and up to 5 days.",
    },
    {
      "text": "After the dough has chilled for at least 2 hours, you can roll it out. Work with one crust at a time, keeping the other in the refrigerator until you’re ready to roll it out. Lightly flour the work surface, rolling pin, and your hands, and sprinkle a little flour on top of the dough. Use gentle-medium force with your rolling pin on the dough—don’t press down too hard on the dough; you’re not mad at it! When rolling dough out, start from the center and work your way out in all directions, turning the dough with your hands as you go. Between passes of the rolling pin, rotate the pie crust and even flip it, to make sure it’s not sticking to your work surface. Sprinkle on a little more flour if it’s sticking; don’t be afraid to use a little more flour. If you notice the dough becoming a lopsided circle as you’re rolling it out, put down the rolling pin and use your hands to help mold the dough back into an even circle. Roll the dough into a very thin 12-inch circle, which is the perfect size to fit a 9-inch pie dish. Your pie dough will be about 1/8 inch thick, which is quite thin. Visible specks of butter and fat in the dough are perfectly normal and expected.",
    },
    {
      "text": "Because your dough is so thin, use your rolling pin to help transfer the pie crust to the pie dish. Carefully roll one end of the circle of dough gently onto the rolling pin, rolling it back towards you, slowly peeling it off the work surface as you go. Pick it up, and carefully roll it back out over the top of the pie dish. It’s helpful to watch how I do it in the video below.",
    },
    {
      "text": "Proceed with the pie per your recipe’s instructions. If your dough requires par-baking, see helpful How to Par-Bake Pie Crust tutorial.",
    },
  ],
  "name": "Homemade Buttery Flaky Pie Crust Recipe",
  "prepTime": 135,
  "rating": 4.8,
  "servings": 2,
  "tags": [
    "pie",
    "american",
    "pie crust",
  ],
}
`;

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Pecan Pie 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "The best pecan pie recipe from my grandma. :)",
  "gallery": [
    "https://www.gimmesomeoven.com/wp-content/uploads/2013/08/Pecan-Pie-8-225x225.jpg",
    "https://www.gimmesomeoven.com/wp-content/uploads/2013/08/Pecan-Pie-8-260x195.jpg",
    "https://www.gimmesomeoven.com/wp-content/uploads/2013/08/Pecan-Pie-8-320x180.jpg",
    "https://www.gimmesomeoven.com/wp-content/uploads/2013/08/Pecan-Pie-8.jpg",
  ],
  "image": "https://www.gimmesomeoven.com/wp-content/uploads/2013/08/Pecan-Pie-8-225x225.jpg",
  "ingredients": [
    {
      "text": "1 unbaked 9-inch pie crust",
    },
    {
      "text": "1 cup granulated (white) sugar",
    },
    {
      "text": "1 cup light corn syrup*",
    },
    {
      "text": "1 teaspoon vanilla extract",
    },
    {
      "text": "1 teaspoon salt",
    },
    {
      "text": "1/4 teaspoon ground cinnamon",
    },
    {
      "text": "4 eggs, whisked",
    },
    {
      "text": "1/2 cup (1 stick) salted butter",
    },
    {
      "text": "4 cups pecan halves, divided",
    },
  ],
  "instructions": [
    {
      "text": "Heat oven to 400°F.",
    },
    {
      "text": "Chill the unbaked pie crust for at least 20 minutes in the refrigerator or freezer.  Use parchment paper or aluminum foil to line the inside of the chilled crust, shaping it gently around the inner edges to form a mold of the crust. Fill the bottom of the crust with pie weights or dried beans.  Bake for 15 minutes. Remove the pie pan from the oven, carefully lift out the foil/parchment and weights, and set them aside.  Decrease the oven heat to 350°F.",
    },
    {
      "text": "In a medium bowl, whisk together sugar, corn syrup, vanilla, salt and cinnamon.  Add eggs and whisk until smooth.  Melt butter in a small saucepan over medium heat for about 4-5 minutes, stirring often, until the butter turns brown and fragrant. Slowly pour butter into the sugar mixture, and whisk to combine. Chop 2 cups of the pecans, then stir them into the butter and sugar mixture until combined.",
    },
    {
      "text": "Pour the entire filling into the prepared pie crust. Line the top of the filling with the remaining pecans (about 2 cups, more or less).",
    },
    {
      "text": "Bake the pie for 40-50 minutes, or until the top is lightly browned. After the first 20 to 25 minutes of baking time, you are welcome to loosely tent a piece of aluminum foil over the top of the whole pie if the crust or pecans seem to be browning too quickly.  The pie will be ready to go once the top has puffed up into a dome (which will sink to become flat again, once the pie has set and cooled).  Transfer the pie to a wire rack and let it rest until it reaches room temperature.",
    },
    {
      "text": "Slice and serve pie at room temperature (or you can refrigerate and chill the pie, if you prefer), garnished with a dollop of whipped cream if desired. Cover and store leftover pie at room temperature for 1-2 days or in the refrigerator for 4-5 days.",
    },
  ],
  "name": "Pecan Pie",
  "prepTime": 60,
  "rating": 4.8,
  "servings": 12,
  "tags": [],
}
`;

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Shepherd’s Pie 1`] = `
{
  "calories": 606,
  "color": "#ffffff",
  "description": "This easy shepherd's pie is a classic comfort food recipe that's hearty, filling, and perfect for the winter season or St. Patrick's Day. Watch the video below to see how I make the recipe!",
  "gallery": [
    "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/Shepherds-Pie-8.jpg",
  ],
  "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/Shepherds-Pie-8.jpg",
  "ingredients": [
    {
      "text": "2 pounds russet potatoes (peeled and quartered)",
    },
    {
      "text": "½ cup milk",
    },
    {
      "text": "¼ cup butter",
    },
    {
      "text": "kosher salt and freshly ground black pepper (to taste)",
    },
    {
      "text": "2  tablespoons avocado oil or olive oil",
    },
    {
      "text": "1  large onion (diced)",
    },
    {
      "text": "3  garlic cloves (minced)",
    },
    {
      "text": "2  large carrots (peeled and diced)",
    },
    {
      "text": "2  celery stalk (diced)",
    },
    {
      "text": "1½ pounds ground lamb or ground beef",
    },
    {
      "text": "1 teaspoon salt",
    },
    {
      "text": "¼ teaspoon pepper",
    },
    {
      "text": "1 cup chicken broth (or if using ground beef use beef broth)",
    },
    {
      "text": "1½ tablespoons tomato paste",
    },
    {
      "text": "1 tablespoon Worcestershire sauce (see notes above)",
    },
    {
      "text": "2 teaspoons fresh rosemary (finely chopped)",
    },
    {
      "text": "1 teaspoon fresh thyme (finely chopped)",
    },
    {
      "text": "1 cup frozen peas",
    },
  ],
  "instructions": [
    {
      "text": "Place the potatoes in a pot on the stove and cover with cold water. Turn the heat to high and bring to a boil. Cook for 13 to 15 minutes, or until the potatoes are soft when pierced with a fork.",
    },
    {
      "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/How-to-Make-Shepherds-Pie-1.jpg",
      "text": "Drain the potatoes in a colander, then return to the pot. Add the milk, butter, salt and pepper and mash until creamy.",
    },
    {
      "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/How-to-Make-Shepherds-Pie.jpg",
      "text": "Preheat your oven to 400°F (200°C). In a large oven-safe saute pan, heat the oil on medium heat. Add the onion and garlic and saute for a minute. Then add the diced carrots, diced celery, and lamb. Cook for 8 to 10 minutes or until the meat is browned. Use your spatula to break up the meat while cooking.",
    },
    {
      "text": "Drain the fat from the pan and add the broth, tomato paste, Worcestershire sauce, rosemary, thyme, salt and pepper. Simmer for about 5 minutes, until the sauce is slightly thickened.",
    },
    {
      "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/How-to-Make-Shepherds-Pie-4.jpg",
      "text": "Add the frozen peas and stir together. Turn off the stove and use the back of your spatula to flatten the meat mixture into a single layer.",
    },
    {
      "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/How-to-Make-Shepherds-Pie-5.jpg",
      "text": "Dollop the mashed potatoes on top of the meat and use a spoon or spatula to spread it flat to the edges. If desired, create texture in the potato topping with a spoon or fork.",
    },
    {
      "image": "https://i2.wp.com/www.downshiftology.com/wp-content/uploads/2019/12/How-to-Make-Shepherds-Pie-6.jpg",
      "text": "Bake the shepherd's pie until slightly golden, about 25 to 30 minutes. You can broil the top for 1 to 2 minutes as well.",
    },
  ],
  "name": "Easy Shepherd's Pie Recipe",
  "prepTime": 60,
  "rating": 4.99,
  "servings": 6,
  "tags": [
    "main course",
    "british",
    "irish",
    "shepherd's pie",
    "shepherd's pie recipe",
  ],
}
`;

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Tuscan Chicken Casserole 1`] = `
{
  "calories": 597,
  "color": "#ffffff",
  "description": "This Tuscan chicken casserole, boneless chicken breasts in a creamy cheese and sun-dried tomato sauce with spinach and mushrooms, requires little enough hands-on time to make for weeknight dinner.",
  "gallery": [
    "https://www.allrecipes.com/thmb/0TPMTnN9PmtWYJngclEUOsom_cs=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc()/8735197_Tuscan-Chicken-Casserole_Yolanda-Gutierrez_4x3-01bc05f1c1514461834e94c8f8c8e3c2.jpg",
  ],
  "image": "https://www.allrecipes.com/thmb/0TPMTnN9PmtWYJngclEUOsom_cs=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc()/8735197_Tuscan-Chicken-Casserole_Yolanda-Gutierrez_4x3-01bc05f1c1514461834e94c8f8c8e3c2.jpg",
  "ingredients": [
    {
      "text": "1 pound skinless, boneless chicken breasts",
    },
    {
      "text": "1 teaspoon Italian seasoning",
    },
    {
      "text": "1/2 teaspoon paprika",
    },
    {
      "text": "1/2 teaspoon onion powder",
    },
    {
      "text": "1/2 teaspoon red pepper flakes",
    },
    {
      "text": "1/2 teaspoon garlic powder",
    },
    {
      "text": "salt and freshly ground black pepper to taste",
    },
    {
      "text": "2 cups fresh spinach",
    },
    {
      "text": "1 (4 ounce) can sliced mushrooms, drained",
    },
    {
      "text": "1/2 cup chopped sun-dried tomatoes",
    },
    {
      "text": "3/4 cup shredded mozzarella cheese",
    },
    {
      "text": "1/4 cup freshly grated parmesan cheese",
    },
  ],
  "instructions": [
    {
      "text": "Preheat the oven to 425 degrees F (220 degrees C). Spray an 8x8-inch baking dish with nonstick cooking spray.",
    },
    {
      "text": "Season both sides of chicken with Italian seasoning, paprika, onion powder, pepper flakes, garlic powder, salt and pepper. Place in the prepared pan. Place spinach over chicken, mushrooms over spinach, and sun-dried tomatoes over mushrooms. Top with mozzarella and Parmesan cheese.",
    },
    {
      "text": "For sauce, beat cream, Parmesan cheese, garlic, cream cheese, chicken bouillon, chicken broth, and oil together in a bowl until just combined. Season with salt and pepper. Pour mixture over chicken.",
    },
    {
      "text": "Bake casserole in the preheated oven until bubbly, chicken is no longer pink, and juices run clear, 20 to 25 minutes.  An instant-read thermometer inserted in the center of chicken should read 165 degrees F (74 degrees C).",
    },
  ],
  "name": "Tuscan Chicken Casserole",
  "prepTime": 40,
  "rating": 4,
  "servings": 4,
  "tags": [
    "dinner",
    "entree",
    "italian",
  ],
}
`;
